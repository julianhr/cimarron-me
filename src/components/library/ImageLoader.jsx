import React from 'react'
import PropTypes from 'prop-types'
import styled from '@emotion/styled'


const Root = styled.figure`
  display: flex;
  align-items: center;
  background: ${props => props.theme.colors.background};
`

const Img = styled.img`
  opacity: 0;
  transition: opacity 100ms linear;
`

class ImageLoader extends React.PureComponent {
  static propTypes = {
    maxWidth: PropTypes.number.isRequired,
    maxHeight: PropTypes.number.isRequired,
    imgSrc: PropTypes.string.isRequired,
    styles: PropTypes.object,
  }

  static defaultProps = {
    styles: {},
  }

  state = {
    isImgLoaded: false,
  }

  refImg = React.createRef()

  handleOnLoad = () => {
    this.setState({ isImgLoaded: true })
  }

  render() {
    const { styles, maxWidth, maxHeight, imgSrc } = this.props
    let imgStyle = { willChange: 'opacity' }

    if (this.state.isImgLoaded) {
      imgStyle = {
        opacity: 1,
        width: this.refImg.current.width,
        height: this.refImg.current.height,
      }
    }

    return (
      <Root
        css={{ ...(styles.root || {}), width: maxWidth, height: maxHeight }}
      >
        <Img
          ref={this.refImg}
          src={imgSrc}
          css={{ ...(styles.img || {}), ...imgStyle }}
          onLoad={this.handleOnLoad}
        />
      </Root>
    )
  }
}

export default ImageLoader
