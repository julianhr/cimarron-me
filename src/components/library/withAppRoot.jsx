import React from 'react'
import { Provider } from 'react-redux'
import { BrowserRouter } from 'react-router-dom'
import { ThemeProvider } from 'emotion-theming'
import { Global } from '@emotion/core'
import 'sanitize.css'

import appStore from '~/reducers'
import globalStyles from '~/styles/globalStyles'
import theme from '~/styles/theme'


function withAppRoot(WrappedComponent, props={}) {
  return () => (
    <Provider store={appStore}>
      <BrowserRouter>
        <ThemeProvider theme={theme}>
          <Global styles={globalStyles} />
          <WrappedComponent {...props} />
        </ThemeProvider>
      </BrowserRouter>
    </Provider>
  )
}

export default withAppRoot
