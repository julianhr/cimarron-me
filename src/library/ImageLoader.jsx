import React from 'react'
import PropTypes from 'prop-types'
import styled from '@emotion/styled'


const Root = styled.figure`
  display: flex;
  align-items: center;
  background: ${props => props.theme.colors.background};
`

class ImageLoader extends React.PureComponent {
  static propTypes = {
    imgSrc: PropTypes.string.isRequired,
    maxHeight: PropTypes.number.isRequired,
    maxWidth: PropTypes.number.isRequired,
    shouldFadeIn: PropTypes.bool,
    styles: PropTypes.object,
  }

  static defaultProps = {
    styles: {},
    shouldFadeIn: true,
  }

  constructor() {
    super()
    this.handleOnLoad = this.handleOnLoad.bind(this)
  }

  state = {
    isImgLoaded: false,
  }

  handleOnLoad() {
    this.setState({ isImgLoaded: true })
  }

  getImgProps() {
    if (this.props.shouldFadeIn) {
      let imgStyle = {
        transition: 'opacity 80ms linear',
        willChange: 'opacity',
        opacity: 0,
      }
  
      if (this.state.isImgLoaded) {
        imgStyle = {
          transition: 'opacity 80ms linear',
          opacity: 1,
          width: '100%',
          height: 'auto',
        }
      }

      return {
        css: [this.props.styles.img, imgStyle],
        onLoad: this.handleOnLoad
      }
    } else {
      return { css: [this.props.styles.img] }
    }
  }

  render() {
    const { styles, maxWidth, maxHeight, imgSrc } = this.props

    return (
      <Root
        css={{ ...(styles.root || {}), width: maxWidth, height: maxHeight }}
      >
        <img
          src={imgSrc}
          {...this.getImgProps()}
        />
      </Root>
    )
  }
}

export default ImageLoader
