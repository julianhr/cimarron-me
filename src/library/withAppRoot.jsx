import React from 'react'
import { Provider } from 'react-redux'
import { BrowserRouter } from 'react-router-dom'
import { ThemeProvider } from 'emotion-theming'
import { Global } from '@emotion/core'
import 'sanitize.css'

import globalStyles from '~/styles/globalStyles'
import theme from '~/styles/theme'


function withAppRoot(WrappedComponent, store) {
  return class extends React.Component {
    render() {
      return (
        <Provider store={store}>
          <BrowserRouter>
            <ThemeProvider theme={theme}>
              <Global styles={globalStyles} />
              <WrappedComponent {...this.props} />
            </ThemeProvider>
          </BrowserRouter>
        </Provider>
      )
    }
  }
}

export default withAppRoot
