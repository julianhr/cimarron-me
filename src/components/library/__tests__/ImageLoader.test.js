import React from 'react'
import { render, mount, shallow } from 'enzyme'
import { toJson } from 'enzyme-to-json'

import ImageLoader from '../ImageLoader'
import withAppRoot from '~/components/library/withAppRoot'
import throwPropErrors from '~/__tests__/__mocks__/throwPropErrorsMock'
import imgSrc from '~/__tests__/__fixtures__/imgSrc'


const testProps = {
  maxWidth: 200,
  maxHeight: 100,
  imgSrc: imgSrc,
  styles: {
    root: {
      color: 'green'
    },
    img: {
      color: 'yellow'
    }
  },
}

beforeAll(() => {
  throwPropErrors()
})

describe('ImageLoader', () => {
  it('matches snapshot with all props', () => {
    const RootedImageLoader = withAppRoot(ImageLoader, testProps)
    const wrapper = render(<RootedImageLoader />)
    expect(wrapper).toMatchSnapshot()
  })

  it('matches snapshot with no styles', () => {
    const props = {...testProps}
    delete props['styles']
    const RootedImageLoader = withAppRoot(ImageLoader, props)
    const wrapper = render(<RootedImageLoader />)
    expect(wrapper).toMatchSnapshot()
  })
  
  it('requires maxWidth prop', () => {
    const props = {...testProps}
    delete props['maxWidth']
    const RootedImageLoader = withAppRoot(ImageLoader, props)
    expect(() => render(<RootedImageLoader />)).toThrow()
  })
  
  it('requires maxHeight prop', () => {
    const props = {...testProps}
    delete props['maxHeight']
    const RootedImageLoader = withAppRoot(ImageLoader, props)
    expect(() => render(<RootedImageLoader />)).toThrow()
  })
  
  it('requires imgSrc prop', () => {
    const props = {...testProps}
    delete props['imgSrc']
    const RootedImageLoader = withAppRoot(ImageLoader, props)
    expect(() => render(<RootedImageLoader />)).toThrow()
  })
  
  it('should trigger onload event when image src is loaded', () => {
    const oldHandle = ImageLoader.prototype.handleOnLoad
    const mockHandleOnLoad = jest.fn()
    ImageLoader.prototype.handleOnLoad = mockHandleOnLoad

    const RootedImageLoader = withAppRoot(ImageLoader, testProps)
    const wrapper = mount(<RootedImageLoader />)
    wrapper.find('img').simulate('load')
    expect(mockHandleOnLoad).toBeCalledTimes(1)
    
    ImageLoader.prototype.handleOnLoad = oldHandle
    wrapper.unmount()
  })
  
  it('matches snapshot after image src is loaded', () => {
    const RootedImageLoader = withAppRoot(ImageLoader, testProps)
    const wrapper = mount(<RootedImageLoader />)
    wrapper.find('img').simulate('load')
    expect(wrapper).toMatchSnapshot()
    wrapper.unmount()
  })
})
