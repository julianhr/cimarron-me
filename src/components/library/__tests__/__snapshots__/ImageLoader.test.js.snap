// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageLoader matches snapshot after image src is loaded 1`] = `
Array [
  <ForwardRef(render)
    styles={
      Object {
        "map": "/*# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbImdsb2JhbFN0eWxlcy5qcyJdLCJuYW1lcyI6W10sIm1hcHBpbmdzIjoiQUFHd0IiLCJmaWxlIjoiZ2xvYmFsU3R5bGVzLmpzIiwic291cmNlc0NvbnRlbnQiOlsiaW1wb3J0IHsgY3NzIH0gZnJvbSAnQGVtb3Rpb24vY29yZSdcbmltcG9ydCB0aGVtZSBmcm9tICcuL3RoZW1lJ1xuXG5jb25zdCBnbG9iYWxTdHlsZXMgPSBjc3NgXG4gIGJvZHkge1xuICAgIGNvbG9yOiAke3RoZW1lLmNvbG9ycy50ZXh0fSxcbiAgICBsaW5lSGVpZ2h0IDEuM2VtO1xuICB9XG5cbiAgcCwgaDEsIGgyLCBoMywgaDQsIGg1LCBoNiwgZmlndXJlIHtcbiAgICBtYXJnaW46IDA7XG4gIH1cblxuICBoMSwgaDIsIGgzLCBoNCwgaDUsIGg2IHtcbiAgICBjb2xvcjogJHt0aGVtZS5jb2xvcnMucHJpbWFyeS5kYXJrfTtcbiAgfVxuXG4gIHAge1xuICAgIHBhZGRpbmc6IC44ZW0gMDtcbiAgfVxuXG4gIGgxIHtcbiAgICBmb250LXNpemU6IDQ0O1xuICB9XG5cbiAgaDIge1xuICAgIGxpbmUtaGVpZ2h0OiAyZW07XG4gICAgZm9udC1zaXplOiAzNDtcbiAgICBmb250LXdlaWdodDogNTAwO1xuICB9XG5cbiAgaDMge1xuICAgIGZvbnQtc2l6ZTogMjQ7XG4gICAgZm9udC13ZWlnaHQ6IDYwMDtcbiAgfVxuXG4gIGg0IHtcbiAgICBsaW5lLWhlaWdodDogMS4zZW07XG4gICAgZm9udC1zaXplOiAxODtcbiAgfVxuXG4gIGg1IHtcbiAgICBmb250LXNpemU6IDEyO1xuICB9XG5cbiAgaDYge1xuICAgIGZvbnQtc2l6ZTogODtcbiAgfVxuXG4gIGNvZGUge1xuICAgIGNvbG9yOiAke3RoZW1lLmNvbG9ycy5jb2RlfTtcbiAgICBmb250LXNpemU6IDE0O1xuICB9XG5gXG5cbmV4cG9ydCBkZWZhdWx0IGdsb2JhbFN0eWxlc1xuIl19 */",
        "name": "1e2xhh9-globalStyles-globalStyles",
        "next": undefined,
        "styles": "body{color:#3D3D3D,lineHeight 1.3em;}p,h1,h2,h3,h4,h5,h6,figure{margin:0;}h1,h2,h3,h4,h5,h6{color:#AB164D;}p{padding:.8em 0;}h1{font-size:44;}h2{line-height:2em;font-size:34;font-weight:500;}h3{font-size:24;font-weight:600;}h4{line-height:1.3em;font-size:18;}h5{font-size:12;}h6{font-size:8;}code{color:#D90000;font-size:14;}label:globalStyles;label:globalStyles;",
      }
    }
  />,
  .emotion-5 {
  color: green;
  width: 200px;
  height: 100px;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #FFF;
  color: green;
  width: 200px;
  height: 100px;
}

.emotion-2 {
  color: yellow;
  opacity: 1;
  width: 0;
  height: 0;
}

.emotion-0 {
  opacity: 0;
  -webkit-transition: opacity 100ms linear;
  transition: opacity 100ms linear;
  color: yellow;
  opacity: 1;
  width: 0;
  height: 0;
}

<Root
    className="emotion-5"
  >
    <figure
      className="emotion-3 emotion-4"
    >
      <Img
        className="emotion-2"
        onLoad={[Function]}
        src="data:image/webp;base64,UklGRlQBAABXRUJQVlA4IEgBAABQCACdASoeABsAPtFco0yoJSMiMBgIAQAaCWwAuzlBfgGnO3BOALUBuQHUAbxt+2ZGxJRCYtBpYkvKE4agKG/XsXzkJxoCkLfy2JgA/v6UMx3P58mCrRxyi2inrTWQtl778n/xgSpf89/Y/+P6MxQ7ElzA7wlfu8XgcKi2XGDX7X/3TJ+Lc2FN9UflwEHsNaDGFm+cp3PSYB0z6nyEc0tkHRfee/M3RlRP686V8jhfSqeGU40Y9cP0betUGiyL6fX4UCGklWdvBvmBYV2/lpyWAvTXYmHBUTOgmRf4mZ8OSrwbxLSnL7H9xwwSzwHcz5vC2vjXJyn7j2hIrhfmENec/2zV6jbNPlXZFZofyE2n8lhCvjL/wDWARF1g0qxg4v8KzSYrKPR0mH4EXAsrO8U81v8+RaHKdg40DXgwNHEX3mNcn7aigAAA"
      >
        <img
          className="emotion-0 emotion-1"
          onLoad={[Function]}
          src="data:image/webp;base64,UklGRlQBAABXRUJQVlA4IEgBAABQCACdASoeABsAPtFco0yoJSMiMBgIAQAaCWwAuzlBfgGnO3BOALUBuQHUAbxt+2ZGxJRCYtBpYkvKE4agKG/XsXzkJxoCkLfy2JgA/v6UMx3P58mCrRxyi2inrTWQtl778n/xgSpf89/Y/+P6MxQ7ElzA7wlfu8XgcKi2XGDX7X/3TJ+Lc2FN9UflwEHsNaDGFm+cp3PSYB0z6nyEc0tkHRfee/M3RlRP686V8jhfSqeGU40Y9cP0betUGiyL6fX4UCGklWdvBvmBYV2/lpyWAvTXYmHBUTOgmRf4mZ8OSrwbxLSnL7H9xwwSzwHcz5vC2vjXJyn7j2hIrhfmENec/2zV6jbNPlXZFZofyE2n8lhCvjL/wDWARF1g0qxg4v8KzSYrKPR0mH4EXAsrO8U81v8+RaHKdg40DXgwNHEX3mNcn7aigAAA"
        />
      </Img>
    </figure>
  </Root>,
]
`;

exports[`ImageLoader matches snapshot with all props 1`] = `
.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #FFF;
  color: green;
  width: 200px;
  height: 100px;
}

.emotion-0 {
  opacity: 0;
  -webkit-transition: opacity 100ms linear;
  transition: opacity 100ms linear;
  color: yellow;
  will-change: opacity;
}

<figure
  class="emotion-2 emotion-3"
>
  <img
    class="emotion-0 emotion-1"
    src="data:image/webp;base64,UklGRlQBAABXRUJQVlA4IEgBAABQCACdASoeABsAPtFco0yoJSMiMBgIAQAaCWwAuzlBfgGnO3BOALUBuQHUAbxt+2ZGxJRCYtBpYkvKE4agKG/XsXzkJxoCkLfy2JgA/v6UMx3P58mCrRxyi2inrTWQtl778n/xgSpf89/Y/+P6MxQ7ElzA7wlfu8XgcKi2XGDX7X/3TJ+Lc2FN9UflwEHsNaDGFm+cp3PSYB0z6nyEc0tkHRfee/M3RlRP686V8jhfSqeGU40Y9cP0betUGiyL6fX4UCGklWdvBvmBYV2/lpyWAvTXYmHBUTOgmRf4mZ8OSrwbxLSnL7H9xwwSzwHcz5vC2vjXJyn7j2hIrhfmENec/2zV6jbNPlXZFZofyE2n8lhCvjL/wDWARF1g0qxg4v8KzSYrKPR0mH4EXAsrO8U81v8+RaHKdg40DXgwNHEX3mNcn7aigAAA"
  />
</figure>
`;

exports[`ImageLoader matches snapshot with no styles 1`] = `
.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #FFF;
  width: 200px;
  height: 100px;
}

.emotion-0 {
  opacity: 0;
  -webkit-transition: opacity 100ms linear;
  transition: opacity 100ms linear;
  will-change: opacity;
}

<figure
  class="emotion-2 emotion-3"
>
  <img
    class="emotion-0 emotion-1"
    src="data:image/webp;base64,UklGRlQBAABXRUJQVlA4IEgBAABQCACdASoeABsAPtFco0yoJSMiMBgIAQAaCWwAuzlBfgGnO3BOALUBuQHUAbxt+2ZGxJRCYtBpYkvKE4agKG/XsXzkJxoCkLfy2JgA/v6UMx3P58mCrRxyi2inrTWQtl778n/xgSpf89/Y/+P6MxQ7ElzA7wlfu8XgcKi2XGDX7X/3TJ+Lc2FN9UflwEHsNaDGFm+cp3PSYB0z6nyEc0tkHRfee/M3RlRP686V8jhfSqeGU40Y9cP0betUGiyL6fX4UCGklWdvBvmBYV2/lpyWAvTXYmHBUTOgmRf4mZ8OSrwbxLSnL7H9xwwSzwHcz5vC2vjXJyn7j2hIrhfmENec/2zV6jbNPlXZFZofyE2n8lhCvjL/wDWARF1g0qxg4v8KzSYrKPR0mH4EXAsrO8U81v8+RaHKdg40DXgwNHEX3mNcn7aigAAA"
  />
</figure>
`;
